airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py

```diff
@@ -472,3 +528,207 @@ class Labels(GoogleAdsStream):
+        for key in ["start_date", "conversion_window_days", "end_date"]:
+            kwargs.pop(key, None)
+        super().__init__(start_date=None, conversion_window_days=0, end_date=None, **kwargs)
+
+    def get_query(self, stream_slice: Mapping[str, Any]) -> str:
+        query = GoogleAds.convert_schema_into_query(
+            schema=self.get_json_schema(),
+            report_name=self.name,
+            from_date=stream_slice.get("start_date"),
+            to_date=stream_slice.get("end_date"),
+            cursor_field=self.cursor_field,
+            # resource type is used for filtering updates only for desirable stream
+            resource_type=stream_slice.get("resource_type"),
+            # limit is mandatory parameter for this stream, also 10k is maximum value
+            limit=self.mandatory_limit,
+        )
+        return query
+
+
+class IncrementalEventsStream(GoogleAdsStream, IncrementalMixin, ABC):
+    """
+    Abstract class used for getting incremental updates based on events returned from ChangeStatus stream
+    """
+
+    def __init__(self, parent_stream: IncrementalGoogleAdsStream, **kwargs):
+        self.parent_stream = parent_stream
+        self.parent_stream_name: str = self.parent_stream.name
+        self.parent_cursor_field: str = self.parent_stream.cursor_field
+        self.parent_sync_mode: SyncMode = SyncMode.incremental
+
+        super().__init__(**kwargs)
+        self.incremental_sieve_logger = cyclic_sieve(self.logger, 10)
+
+        self._state = {self.parent_stream_name: {customer.id: None for customer in self.customers}}
+
+    @property
+    @abstractmethod
+    def id_field(self) -> str:
+        "Name of field used for getting records by id"
+        pass
+
+    @property
+    @abstractmethod
+    def parent_id_field(self) -> str:
+        "Field name of id from parent record"
+        pass
+
+    @property
+    @abstractmethod
+    def resource_type(self) -> str:
+        "Resource type used for filtering parent records"
+        pass
+
+    @property
+    def state(self) -> MutableMapping[str, Any]:
+        return self._state
+
+    @state.setter
+    def state(self, value: MutableMapping[str, Any]):
+        self._state.update(value)
+        self.parent_stream.state = self._state.get(self.parent_stream_name, {})
+
+    def current_state(self, customer_id, default=None):
+        return self.parent_stream.current_state(customer_id, default)
+
+    def stream_slices(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Iterable[Optional[MutableMapping[str, any]]]:
+        """
+        If state exists read updates from parent stream otherwise return slices with only customer id to sync all records for stream
+        """
+        if stream_state:
+            slices_generator = self.read_parent_stream(self.parent_sync_mode, self.parent_cursor_field, stream_state)
+            yield from slices_generator
+        else:
+            for customer in self.customers:
```

Comment:
why are we keeping the IDs of all deleted records in the stream slice, instead of doing something like this in `read_records` : 

```python
for record in parent.read_records():
  if parent_record.get("change_status.resource_status") == "REMOVED":
    yield {self.id_field: parent_record.get(id), "deleted_at": parent_record.get(updated_at)}
```
? 

It feels like structuring this this way inlined in `read_records` would be a lot simpler to understand

Reply:
During stream slicing, we collect ids of updated records and then request them in one batch. To avoid reading parent two times during slicing and reading deleting records we save results in slicing.